pkg/executor/join/anti_semi_join_probe.go

```diff
@@ -0,0 +1,337 @@
+		}
+		a.rowIter.next()
+	}
+	err := checkSQLKiller(sqlKiller, "killedDuringProbe")
+	if err != nil {
+		joinResult.err = err
+		return joinResult
+	}
+	if a.nextCachedBuildRowIndex > 0 {
+		a.batchConstructBuildRows(joinResult.chk, 0, false)
+	}
+	return joinResult
+}
+
+func (a *antiSemiJoinProbe) ResetProbe() {
+	a.rowIter = nil
+	a.baseJoinProbe.ResetProbe()
+}
+
+func (a *antiSemiJoinProbe) resetProbeState() {
+	a.baseSemiJoin.resetProbeState()
+
+	if !a.isLeftSideBuild {
+		if a.ctx.spillHelper.isSpillTriggered() {
+			for _, idx := range a.spilledIdx {
+				// We see rows that have be spilled as matched rows
+				a.isMatchedRows[idx] = true
+			}
+		}
+	}
+
+	a.outputProbeRowsSelected = false
+}
+
+func (a *antiSemiJoinProbe) SetChunkForProbe(chk *chunk.Chunk) (err error) {
+	err = a.baseJoinProbe.SetChunkForProbe(chk)
+	if err != nil {
+		return err
+	}
+
+	a.resetProbeState()
+	return nil
+}
+
+func (a *antiSemiJoinProbe) SetRestoredChunkForProbe(chk *chunk.Chunk) error {
+	err := a.baseJoinProbe.SetRestoredChunkForProbe(chk)
+	if err != nil {
+		return err
+	}
+
+	a.resetProbeState()
+	return nil
+}
+
+func (a *antiSemiJoinProbe) Probe(joinResult *hashjoinWorkerResult, sqlKiller *sqlkiller.SQLKiller) (ok bool, _ *hashjoinWorkerResult) {
+	if joinResult.chk.IsFull() {
+		return true, joinResult
+	}
+
+	joinedChk, remainCap, err := a.prepareForProbe(joinResult.chk)
+	if err != nil {
+		joinResult.err = err
+		return false, joinResult
+	}
+
+	hasOtherCondition := a.ctx.hasOtherCondition()
+	if a.isLeftSideBuild {
+		if hasOtherCondition {
+			err = a.probeForLeftSideBuildHasOtherCondition(joinedChk, sqlKiller)
+		} else {
+			err = a.probeForLeftSideBuildNoOtherCondition(sqlKiller)
+		}
+	} else {
+		if hasOtherCondition {
+			err = a.probeForRightSideBuildHasOtherCondition(joinResult.chk, joinedChk, remainCap, sqlKiller)
+		} else {
+			err = a.probeForRightSideBuildNoOtherCondition(joinResult.chk, remainCap, sqlKiller)
+		}
+	}
+	if err != nil {
+		joinResult.err = err
+		return false, joinResult
+	}
+	return true, joinResult
+}
+
+func (a *antiSemiJoinProbe) probeForLeftSideBuildHasOtherCondition(joinedChk *chunk.Chunk, sqlKiller *sqlkiller.SQLKiller) (err error) {
+	err = a.concatenateProbeAndBuildRows(joinedChk, sqlKiller, false)
+	if err != nil {
+		return err
+	}
+
+	if a.unFinishedProbeRowIdxQueue.IsEmpty() {
+		// To avoid `Previous chunk is not probed yet` error
+		a.currentProbeRow = a.chunkRows
+	}
+
+	meta := a.ctx.hashTableMeta
+	if joinedChk.NumRows() > 0 {
+		a.selected, a.isNulls, err = expression.VecEvalBool(a.ctx.SessCtx.GetExprCtx().GetEvalCtx(), a.ctx.SessCtx.GetSessionVars().EnableVectorizedExpression, a.ctx.OtherCondition, joinedChk, a.selected, a.isNulls)
+		if err != nil {
+			return err
+		}
+
+		for index, result := range a.selected {
+			if result || a.isNulls[index] {
+				candidateRow := unsafe.Pointer(&a.rowIndexInfos[index].buildRowStart)
+				if !meta.isCurrentRowUsedWithAtomic(candidateRow) {
```

Comment:
since matchMultiBuildRows already check this, I think we can just set the flag.

Reply:
> since matchMultiBuildRows already check this, I think we can just set the flag.

done